File: src/main/java/com/zanoccio/axirassa/webapp/services/validation/CustomValidationDecorator.java

package com.zanoccio.axirassa.webapp.services.validation;

import org.apache.tapestry5.BaseValidationDecorator;
import org.apache.tapestry5.CSSClassConstants;
import org.apache.tapestry5.Field;
import org.apache.tapestry5.MarkupWriter;
import org.apache.tapestry5.ValidationTracker;
import org.apache.tapestry5.dom.Element;
import org.apache.tapestry5.services.Environment;

/**
 * 
 * Based on:
 * http://piraya-blog.blogspot.com/2009/06/creating-custom-validationdecorator
 * -for.html
 * 
 * @author wiktor
 * 
 */
public class CustomValidationDecorator extends BaseValidationDecorator {
	private final MarkupWriter markupwriter;
	private final Environment environment;


	public CustomValidationDecorator(final Environment environment, final MarkupWriter markupwriter) {
		this.environment = environment;
		this.markupwriter = markupwriter;
	}


	@Override
	public void insideLabel(final Field field, final Element element) {
		if (field != null) {
			if (inError(field))
				element.addClassName(CSSClassConstants.ERROR);
			if (field.isRequired())
				element.addClassName("required");
		}
	}


	@Override
	public void insideField(final Field field) {
		if (field != null && inError(field)) {
			markupwriter.getElement().addClassName(CSSClassConstants.ERROR);
		}
	}


	@Override
	public void afterField(final Field field) {
		if (field != null) {
			ValidationTracker tracker = environment.peekRequired(ValidationTracker.class);
			if (tracker.inError(field)) {
				String error = tracker.getError(field);
				if (error != null && error.length() > 0) {
					markupwriter.element("span", "class", CSSClassConstants.ERROR);
					markupwriter.write(error);
					markupwriter.end();
				}
			}
		}
	}


	private boolean inError(final Field field) {
		System.out.println("Checking for errir in field" + field.toString());

		ValidationTracker tracker = environment.peekRequired(ValidationTracker.class);
		return tracker.inError(field);
	}
}

File: src/main/java/axirassa/webapp/services/validation/CustomValidationDecorator.scala

package axirassa.webapp.services.validation

import org.apache.tapestry5.BaseValidationDecorator
import org.apache.tapestry5.CSSClassConstants
import org.apache.tapestry5.Field
import org.apache.tapestry5.MarkupWriter
import org.apache.tapestry5.ValidationTracker
import org.apache.tapestry5.dom.Element
import org.apache.tapestry5.services.Environment

/**
 * Validations are annoying because there are both client and server side
 * components, each necessarily implemented in Java vs JavaScript.
 * 
 * Based on: http://jumpstart.doublenegative.com.au/jumpstart/examples/input/
 * novalidationbubbles1
 * 
 * @author wiktor
 * 
 */
class CustomValidationDecorator(environment : Environment, markupwriter : MarkupWriter) extends BaseValidationDecorator {
	override
	def insideLabel(field : Field, element : Element) {
		if (field == null)
			return

		if (field.isRequired()) {
			element.addClassName("required-label")
			element.getContainer.addClassName("required-label-c")
		}

		if (inError(field)) {
			element.addClassName("error-label")
			element.getContainer.addClassName("error-label-c")
		}
	}


	override
	def insideField(field : Field) {
		if (field != null && inError(field)) {
			markupwriter.getElement.addClassName(CSSClassConstants.ERROR)
		}
	}


	override
	def afterField( field : Field) {
		if (field.isRequired()) {
			getElement.addClassName("required-field")
			getElement.getContainer.addClassName("required-field-c")
		}

		if (inError(field)) {
			getElement.addClassName("error-field")
			getElement.getContainer.addClassName("error-field-c")
		}
	}

	private def inError(field : Field) = getTracker.inError(field)
	private def getTracker()  = environment.peekRequired(classOf[ValidationTracker])
	private def getElement = markupwriter.getElement
}